File: examples3d/debug_pop3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    let ground_size = 10.0;
    let ground_height = 10.0;

    for _ in 0..1 {
        let rigid_body =
            RigidBodyBuilder::fixed().translation(Vector::new(0.0, -ground_height, 0.0));
        let collider = ColliderBuilder::cuboid(ground_size, ground_height, ground_size);
        let (_handle, _) = world.insert(rigid_body, collider);
    }

    // Build the dynamic box rigid body.
    for _ in 0..1 {
        let rigid_body = RigidBodyBuilder::dynamic()
            // .translation(Vector::new(0.0, 0.1, 0.0))
            // .rotation(Vector::new(0.8, 0.2, 0.1))
            .can_sleep(false);
        let collider = ColliderBuilder::cuboid(1.0, 1.0, 1.0);
        let (_handle, _) = world.insert(rigid_body, collider.clone());
    }

    /*
     * Set up the testbed.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec3::new(10.0, 10.0, 10.0), Vec3::ZERO);
}

File: python/rapier-testbed/rapier_testbed/examples3/debug_pop3.py
"""Port of ``examples3d/debug_pop3.rs``.

Tiny scene: one dynamic cube starts inside a fixed cuboid, and pops out
when the simulation starts.
"""
from __future__ import annotations

import rapier3d as rp

from .._registry import register

CATEGORY = "Debug"
NAME = "Pop"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 10.0
    ground_height = 10.0

    body = rp.RigidBody.fixed().translation((0.0, -ground_height, 0.0))
    handle = bodies.insert(body)
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height, ground_size), handle, bodies
    )

    body = rp.RigidBody.dynamic().can_sleep(False)
    handle = bodies.insert(body)
    colliders.insert_with_parent(rp.Collider.cuboid(1.0, 1.0, 1.0), handle, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((10.0, 10.0, 10.0), (0.0, 0.0, 0.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
